### DISCUSSION TITLE
What are the param(eter)s?

### DISCUSSION THREAD

**Discussion by JohnDelfosse** [unl-nimbus-lab/pymavswarm#73] [Q&A] [answered]
What are the parameters/what do they do? I understand they hold a key/value pair to be sent to be read/written, but what are these keys and values that they hold, and what would they do? They seem similar in design to messages, and they have several similar checks/retries implemented, but I imagine they have a completely different purpose. What are they parameters to, and why are they being sent/tried to be resent like messages? They don't seem to be like parameters to a function or something.

**Comment by evan-palmer** [comment, marked as answer]
`Parameters` in the `pymavswarm` context provide an interface for getting and setting the parameters of a drone (or some other robot). Some example parameters that may be set or read can be observed [here](https://ardupilot.org/copter/docs/parameters.html). To answer your question regarding why they are wrapped the way that they are, there are a couple of reasons:
1. It is valuable to ensure that, when a request is made to set the value of a parameter on an agent, that parameter is in fact set and set to the correct value
2. It is also useful to confirm that a parameter is read when a request is made for a parameter value. You can think of this like a service.
3. The way that other systems handle this (e.g., MissionPlanner) is all parameters on an agent are loaded at initialization and set at bulk. Because we are working with an unknown number of robotic systems, loading all parameters at initialization would be a prohibitively expensive task
4. We want to bind events to the success/failure of a parameter request for other systems to use for acknowledgement

**Reply by JohnDelfosse** [reply]
These parameters still seem a lot like messages to me. My current understanding is: parameters are basically similar to messages, but instead of communicating instructions, it communicates properties to/from/about the agent. Is this sort of correct?
(I may also be getting this confused with "states" but I guess one thing at a time)

**Reply by evan-palmer** [reply]
They are still a type of message. In the upcoming version (`v0.0.5`) parameters will actually be included in the `messages/` directory along with other commands. The intention is to interact specifically with the parameter set stored internally on an agent.
